File: big-data-core/src/main/scala/Config/PlotGenerationConf/Execution/DTO/BarDTO.scala
package Config.PlotGenerationConf.Execution.DTO

import upickle.default.{ReadWriter, macroRW}
import upickle.implicits.key

case class BarDTOAxisComponent(
  @key("name") name: String,
  @key("format") format: Option[String]
)
object BarDTOAxisComponent { implicit val rw: ReadWriter[BarDTOAxisComponent] = macroRW }

case class BarDTOAxis(
  @key("x") x: BarDTOAxisComponent,
  @key("y") y: BarDTOAxisComponent
)
object BarDTOAxis { implicit val rw: ReadWriter[BarDTOAxis] = macroRW }

case class BarDTOSrc(
  @key("path") path: String,
  @key("axis") axis: BarDTOAxis
)
object BarDTOSrc { implicit val rw: ReadWriter[BarDTOSrc] = macroRW }

case class BarDTODest(
  @key("path") path: String,
  @key("filename") filename: String,
  @key("export_png") exportPng: Boolean
)
object BarDTODest { implicit val rw: ReadWriter[BarDTODest] = macroRW }

case class BarDTOLetteringInsideInfoComponentBuildComponent(
  @key("text_before") textBefore: String,
  @key("name") name: String,
  @key("text_after") textAfter: String
)
object BarDTOLetteringInsideInfoComponentBuildComponent { implicit val rw: ReadWriter[BarDTOLetteringInsideInfoComponentBuildComponent] = macroRW }

case class BarDTOLetteringInsideInfoComponent(
  @key("label") label: String,
  @key("build") build: List[BarDTOLetteringInsideInfoComponentBuildComponent]
)
object BarDTOLetteringInsideInfoComponent { implicit val rw: ReadWriter[BarDTOLetteringInsideInfoComponent] = macroRW }

case class BarDTOLettering(
  @key("title") title: String,
  @key("subtitle") subtitle: String,
  @key("x_label") xLabel: String,
  @key("y_label") yLabel: String,
  @key("inside_info") insideInfo: Option[List[BarDTOLetteringInsideInfoComponent]]
)
object BarDTOLettering { implicit val rw: ReadWriter[BarDTOLettering] = macroRW }

case class BarDTOFigure(
  @key("inverted_horizontal_axis") invertedHorizontalAxis: Boolean,
  @key("threshold_limit_max_min") thresholdLimitMaxMin: Float,
  @key("threshold_perc_limit_outside_text") thresholdPercLimitOutsideText: Float,
  @key("range_margin_perc") rangeMarginPerc: Float,
  @key("color") color: String
)
object BarDTOFigure { implicit val rw: ReadWriter[BarDTOFigure] = macroRW }

case class BarDTOMargin(
  @key("left") left: Float,
  @key("right") right: Float,
  @key("top") top: Float,
  @key("bottom") bottom: Float
)
object BarDTOMargin { implicit val rw: ReadWriter[BarDTOMargin] = macroRW }

case class BarDTOStyle(
  @key("lettering") lettering: BarDTOLettering,
  @key("figure") figure: BarDTOFigure,
  @key("margin") margin: Option[BarDTOMargin]
)
object BarDTOStyle { implicit val rw: ReadWriter[BarDTOStyle] = macroRW }

case class BarDTO (
  @key("src") src: BarDTOSrc,
  @key("dest") dest: BarDTODest,
  @key("style") style: BarDTOStyle
)
object BarDTO { implicit val rw: ReadWriter[BarDTO] = macroRW }

File: auto-plot/App/Api/DTOS/bar_dto.py
"""
DTOs for Bar endpoint.

Defines the request payload model (flask-restx) expected by the bar API.

.. module:: App.Api.DTOS.bar_dto
"""

from flask_restx import fields, Namespace


class BarDTO:
    """Container for the Bar endpoint request model.

    Exposes `post_input` which is a flask-restx model describing the expected JSON
    payload for POST requests to the bar endpoint.

    :param ns: The flask-restx Namespace used to register models and routes.
    :type ns: flask_restx.Namespace
    """
    def __init__(self, ns: Namespace):
        self.post_input = _create_input_post_dto(ns)


def _create_input_post_dto(ns: Namespace):
    """Create and return the flask-restx model describing the POST input payload.

    :param ns: The flask-restx Namespace used to register the model.
    :type ns: flask_restx.Namespace
    :returns: A flask-restx model describing the bar POST payload.
    """
    return ns.model('BarInput', {
        'src': fields.Nested(ns.model('BarSrc', {
            'path': fields.String(required=True, description="Relative route to data"),
            'axis': fields.Nested(ns.model('BarSrcAxis', {
                'x': fields.Nested(ns.model('BarSrcAxisX', {
                    'name': fields.String(reqired=True, description="X Column name")
                }), required=True),
                'y': fields.Nested(ns.model('BarSrcAxisY', {
                    'name': fields.String(reqired=True, description="Y Column name")
                }), required=True)
            }), required=True)
        })),
        'dest': fields.Nested(ns.model('BarDest', {
            'path': fields.String(required=True, description="Relative output route"),
            'filename': fields.String(required=True, description="Filename of the chart"),
            'export_png': fields.Boolean(required=True, description="Export png"),
        }), required=True),
        'style': fields.Nested(ns.model('BarStyle', {
            'lettering': fields.Nested(ns.model('BarStyleLettering', {
                'title': fields.String(required=True, description="Title of the chart"),
                'subtitle': fields.String(required=False, description="Subtitle of the chart"),
                'x_label': fields.String(required=True, description="X axis label"),
                'y_label': fields.String(required=True, description="Y axis label"),
                "inside_info": fields.List(fields.Nested(ns.model('BarStyleLetteringInsideInfoItem', {
                    'label': fields.String(required=True, description="Label for information"),
                    'build': fields.List(fields.Nested(ns.model('BarStyleLetteringInsideInfoItemBuildItem', {
                        'text_before': fields.String(required=True, description="Text before"),
                        'name': fields.String(required=True, description="Name of the column"),
                        'text_after': fields.String(required=True, description="Text after"),
                    })), required=True)
                })), required=False),
            }), required=True),
            'figure': fields.Nested(ns.model('BarStyleFigure', {
                'inverted_horizontal_axis': fields.Boolean(required=True, description="Inverted horizontal axis"),
                'threshold_limit_max_min': fields.Float(required=True,
                                                        description="Threshold limit of extension between the maximum and minimum value to zoom the chart"),
                'threshold_perc_limit_outside_text': fields.Float(required=True,
                                                                  description="Threshold percentage limit for putting text outside the bars"),
                'range_margin_perc': fields.Float(required=True, description="Percentage margin of the value ranges"),
                'color': fields.String(required=True, description="Figure color"),
            }), required=True),
            'margin': fields.Nested(ns.model('BarStyleMargin', {
                'left': fields.Float(required=True, description="Margin left"),
                'right': fields.Float(required=True, description="Margin right"),
                'top': fields.Float(required=True, description="Margin top"),
                'bottom': fields.Float(required=True, description="Margin bottom")
            }), required=False)
        }), required=True)
    })
